cpu/esp8266/periph/rtt.c

```diff
@@ -0,0 +1,201 @@
+
+extern volatile frc2_struct_t frc2;
+
+typedef struct {
+    uint32_t alarm;         /**< alarm */
+    rtt_cb_t alarm_cb;      /**< alarm callback */
+    rtt_cb_t overflow_cb;   /**< overflow callback */
+    void *alarm_arg;        /**< argument for alarm callback */
+    void *overflow_arg;     /**< argument for overflow callback */
+} rtt_config_t;
+
+static rtt_config_t rtt_config = { };
+
+static uint32_t RTC_BSS_ATTR _rtt_counter_saved;
+static uint32_t RTC_BSS_ATTR _rtc_counter_saved;
+
+extern uint32_t pm_rtc_clock_cali_proc(void);
+extern uint32_t pm_rtc2usec(uint32_t rtc_cycles, uint32_t period);
+
+void IRAM rtt_cb(void *arg)
+{
+    if (rtt_config.alarm && rtt_config.alarm_cb) {
+        /* reset the alarm first */
+        rtt_cb_t alarm_cb = rtt_config.alarm_cb;
+        void * alarm_arg = rtt_config.alarm_arg;
+        rtt_clear_alarm();
+        frc2.alarm = 0;
+        /* call the alarm handler afterwards */
+        alarm_cb(alarm_arg);
+    }
+    else if (rtt_config.overflow_cb) {
+        rtt_config.overflow_cb(rtt_config.overflow_arg);
+    }
+}
+
+void rtt_init(void)
+{
+    DEBUG("%s saved rtt=%u rtc=%u\n",
+          __func__, _rtt_counter_saved, _rtc_counter_saved);
+
+    frc2.ctrl.clk_div = FRC2_CLK_DIV_256;
+    frc2.ctrl.reload = 0;
+    frc2.ctrl.intr_hold = 0;
+    frc2.ctrl.enable = 1;
+
+    if (_rtt_counter_saved || _rtc_counter_saved) {
+        /* if in init after power on, restore the RTT counter value */
+        uint32_t rtc_diff = RTC.COUNTER - _rtc_counter_saved;
+        uint32_t rtc_diff_us = pm_rtc2usec(rtc_diff, pm_rtc_clock_cali_proc());
+        uint32_t rtt_diff = RTT_US_TO_TICKS(rtc_diff_us);
+
+        frc2.load = _rtt_counter_saved + rtt_diff;
+
+        DEBUG("%s rtc_diff=%u rtt_diff=%u load=%u\n", __func__,
+              rtc_diff, rtt_diff, _rtt_counter_saved + rtt_diff);
+    }
+    else {
+        frc2.load = 0;
+        DEBUG("%s after power on\n", __func__);
+    }
+
+    /* emulate overflow interrupt */
+    frc2.alarm = 0;
+
+    ets_isr_attach (ETS_FRC2_INUM, rtt_cb, NULL);
+    ets_isr_unmask (BIT(ETS_FRC2_INUM));
+    DPORT.INT_ENABLE |= DPORT_INT_ENABLE_FRC2;
+}
+
+void rtt_poweron(void)
+{
+    /* power on simply reactivates the FRC2 counter */
+    frc2.ctrl.enable = 1;
+}
+
+void rtt_poweroff(void)
+{
+    /* power off simply deactivates the FRC2 counter */
+    frc2.ctrl.enable = 0;
+}
+
+void rtt_set_overflow_cb(rtt_cb_t cb, void *arg)
+{
+    /* there is no overflow interrupt, we emulate */
+    rtt_config.overflow_cb = cb;
+    rtt_config.overflow_arg = arg;
+}
+
+void rtt_clear_overflow_cb(void)
+{
+    /* there is no overflow interrupt, we emulate */
+    rtt_config.overflow_cb = NULL;
+    rtt_config.overflow_arg = NULL;
+}
+
+uint32_t rtt_get_counter(void)
+{
+    return frc2.count;
+}
+
+void rtt_set_counter(uint32_t counter)
+{
+    frc2.load = counter;
+}
+
+void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
+{
+    rtt_config.alarm = alarm;
```

Comment:
`alarm` can happen after the wrap around, so the overflow callback should be triggered before.

Reply:
Indeed. It requires also to change the interrupt handling in `_rtt_cb`.

BTW, configured alarms will not survive deep sleep and reboots. Otherwise the whole `rtt_config_t` structure would have to be placed in RTC RAM which has only 256 bytes. But since the system is restarted after a reboot or deep sleep, preserving configured interrupts makes no sense.

Reply:
OK, seems to work now:
```
> rtt setalarm 1000
> rtt set 4294967290
> The alarm rang

> rtt setoverflow
> rtt setalarm 1000
> rtt set 4294967290
> RTT overflow
> The alarm rang

> rtt setalarm 4294967295
> rtt set 4294967290
> The alarm rang
> RTT overflow

> rtt set 4294967290
> RTT overflow

> rtt setalarm 2000000000
> rtt set 4294967290
> RTT overflow

> rtt set 1999999990
> The alarm rang
```
Because of the interrupt latency, the only case that cannot be handled with this emulated overflow interrupt is when the alarm is set to 0.

Reply:
> Because of the interrupt latency, the only case that cannot be handled with this emulated overflow interrupt is when the alarm is set to 0.

But if you can set an overflow 'alarm' you can also set an alarm at 0, they should happen at the same time.

Reply:
@benpicco Thanks for your suggestions. I had simpler solution already
```
> rtt setoverflow
> rtt setalarm 0
> rtt set 4294967290
The alarm rang
RTT overflow

> rtt set 4294967290
RTT overflow
```

Reply:
My problem was that I tried to test the current counter value, which of course can be larger than the configured alarm value due to the interrupt latency.

So the solution was not to test the actual counter value but the configured alarm value. Since `_rtt_cb` is always called for the actually set alarm value, the value in the `frc2.alarm` register is always exactly the relevant alarm.